UniversalAlgebra/Complexity/FiniteCSP.lagda

```diff
@@ -0,0 +1,138 @@
+
+open import Agda.Builtin.Equality using ( _≡_ ; refl )
+open import Data.Bool             using ( Bool ; true ; false ; _∧_)
+open import Data.Fin.Base         using ( Fin ; toℕ ; fromℕ ; raise)
+open import Data.Vec              using ( Vec ; [] ; tabulate ; lookup ; _∷_ ; map )
+open import Data.Vec.Relation.Unary.All using ( All )
+open import Data.Nat              using ( ℕ ; zero ; suc ; _+_ )
+
+open import Agda.Primitive   using ( _⊔_ ; lsuc ; Level) renaming ( Set to Type )
+open import Function.Base    using ( _∘_ )
+open import Relation.Binary  using ( DecSetoid)
+
+private variable
+ α ℓ : Level
+
+\end{code}
+
+#### Constraints
+
+We start with a vector (of length n, say) of variables.  For simplicity, we'll use the natural
+numbers for variable symbols, so
+
+V : Vec ℕ n
+V = [0 1 2 ... n-1]
+
+We can use the following range function to construct the vector of variables.
+
+\begin{code}
+
+range : (n : ℕ) → Vec ℕ n
+range n = tabulate toℕ
+-- `range n` is 0 ∷ 1 ∷ 2 ∷ … ∷ n-1 ∷ []
+
+\end{code}
+
+Let `nvar` denote the number of variables.
+
+A *constraint* consists of
+
+1. a natural number `∣s∣` denoting the number of variables in the "scope" of the constraint;
+
+2. a scope vector,  `s : Vec (Fin nvar) ∣s∣` , where `s i` is `k` if `k` is the i-th variable
+   in the scope of the constraint.
+
+3. a constraint relation, rel, which is a collection of functions mapping (indices of)
+   variables in the scope to elements of the domain.
+
+To summarize, a constraint is a triple (∣s∣ , s , rel), where
+* ∣s∣ is the number of variables in scope s.
+* s is the scope function: s i ≡ v iff v is the i-th variable in scope s.
+* rel is the contraint relation: a collection of functions mapping indices
+  of scope variables to elements in the domain.
+
+\begin{code}
+
+open DecSetoid
+open Fin renaming (zero to fzer ; suc to fsuc)
+
+record Constraint (nvar : ℕ) (dom : Vec (DecSetoid α ℓ) nvar) : Type α where
+ field
+  ∣s∣ : Fin nvar                     -- The "number" of variables involved in the constraint.
+  s : Vec (Fin nvar) (toℕ ∣s∣)        -- Vec of variables involved in the constraint.
+  rel : ((i : Fin (toℕ ∣s∣)) → Carrier ((lookup dom) (lookup s i))) →  Bool
+         -- `rel f` returns true iff the function f belongs to the relation
+
+ satisfies : (∀ i → Carrier ((lookup dom) i)) → Bool         -- An assignment f of values to variables
+ satisfies f = rel (λ (i : Fin (toℕ ∣s∣)) → f (lookup s i))  -- *satisfies* the constraint provided
+                                                             -- the function f, evaluated at each variable
+                                                             -- in the scope, belongs to the relation rel.
+
+
+-- utility functions --
+foldleft : ∀ {α β} {A : Type α} {B : Type β} {m} →
+           (B → A → B) → B → Vec A m → B
+foldleft _⊕_ b []       = b
+foldleft _⊕_ b (x ∷ xs) = foldleft _⊕_ (b ⊕ x) xs
+-- cf. stdlib's foldl, which seems harder to use than this one
```

Comment:
Harder to _use_ ?? I use it all the time, with no problem!

Reply:
As you know, the `foldl` of the standard library uses dependent types to allow even type B (the type of the left operand in the fold operation) to vary with each index of the vector  That's very impressive for its flexibility, and is more than what is needed here.  For this simple use case, I didn't think the added complexity was worth taking the time to figure out... at least not at this stage.  But maybe I was being a bit lazy and wanting the nice and simple interface to foldleft that I'm comfortable with (e.g., from Scala).  In any case, I'll learn how to use the stdlib's `foldl`.

Reply:
In the rest of the library, there would be a `foldl'` that would be exactly your `foldleft`. So that appears to be an oversight. Something easy to submit to stdlib. Do you want to do it, or would you rather that I did?

Reply:
Go for it!

I wouldn't have recognized that it's missing (...if indeed that's what you're saying).

Reply:
Actually, don't bother adding foldl' on my account. I now see that one need only pass in `\ _ -> B` as the "dependent" type to get the simpler version, so you're right...not hard at all.